python/tvm/driver/tvmc/compiler.py

```diff
@@ -0,0 +1,305 @@
+        needs to be generated.
+    tuning_records: str, optional
+        Name of the file produced by the tuning to be used during
+        compilation.
+    alter_layout: str, optional
+        The layout to convert the graph to. Note, the convert layout
+        pass doesn't currently guarantee the whole of the graph will
+        be converted to the chosen layout.
+
+    Returns
+    -------
+    graph : str
+        A JSON-serialized TVM execution graph.
+    lib : tvm.module.Module
+        A TVM module containing the compiled functions.
+    params : dict
+        The parameters (weights) for the TVM module.
+    dumps : dict
+        Dictionary containing the dumps specified.
+
+    """
+
+    if alter_layout:
+        mod = common.convert_graph_layout(mod, alter_layout)
+
+    if os.path.exists(str(target)):
+        with open(target) as target_file:
+            logging.info("using target input from file: %s", target)
+            target = "".join(target_file.readlines())
+
+    # TODO: We don't have an API to collect a list of supported
+    #       targets yet. (@leandron)
+    logging.debug("creating target from input: %s", target)
+    tvm_target = tvm.target.create(target)
+    target_host = target_host or ""
+
+    if tuning_records:
+        logging.debug("tuning records file provided: %s", tuning_records)
+        with autotvm.apply_history_best(tuning_records):
+            with tvm.transform.PassContext(opt_level=3):
+                logging.debug("building relay graph with tuning records")
+                graph_module = relay.build(mod, tvm_target, params=params, target_host=tvm_target)
+    else:
+        with tvm.transform.PassContext(opt_level=3):
+            logging.debug("building relay graph (no tuning records provided)")
+            graph_module = relay.build(mod, tvm_target, params=params, target_host=tvm_target)
+
+    # Generate output dump files with sources
+    dump_sources = dump_sources or []
+    dumps = {}
+    for source_type in dump_sources:
+        lib = graph_module.get_lib()
+        # TODO lib.get_source call have inconsistent behavior for unsupported
+        #      formats (@leandron).
+        source = str(mod) if source_type == "relay" else lib.get_source(source_type)
+        dumps[source_type] = source
+
+    return graph_module.get_json(), graph_module.get_lib(), graph_module.get_params(), dumps
+
+
+def save_module(module_path, graph, lib, params, cross=None):
+    """
+    Create a tarball containing the generated TVM graph,
+    exported library and parameters
+
+    Parameters
+    ----------
+    module_path : str
+        path to the target tar.gz file to be created,
+        including the file name
+    graph : str
+        A JSON-serialized TVM execution graph.
+    lib : tvm.module.Module
+        A TVM module containing the compiled functions.
+    params : dict
+        The parameters (weights) for the TVM module.
+    cross : Union[str, Callable[[str, str, Optional[str]], None]]
```

Comment:
Please be consistent in the type annotation style. For example, you used numpy-style (e.g., `dict, optional`) in other functions like `compile_relay`.

Reply:
I'll standardise on numpy format.